File: src/leetcode/third/Code_200_NumberOfIslands.java
package leetcode.third;

/**
 * 岛屿数量
 * 
 * @since 2020-7-31 17:03 - 17:17
 */
public class Code_200_NumberOfIslands {
    int numIslands(int[][] grid) {
        if (grid.length == 0)
            return 0;
        int res = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1) {
                    helper(grid, i, j);
                    res++;
                }
            }
        }
        return res;
    }

    private void helper(int[][] grid, int i, int j) {
        if (i >= grid.length || i < 0 || j >= grid[0].length || j < 0 || grid[i][j] != 1) {
            return;
        }
        grid[i][j] = 2;
        helper(grid, i - 1, j);
        helper(grid, i, j - 1);
        helper(grid, i + 1, j);
        helper(grid, i, j + 1);
    }

    public static void main(String[] args) {
        Code_200_NumberOfIslands c = new Code_200_NumberOfIslands();
        // System.out.println(c.numIslands(new char[][] { { '1', '1', '0', '0', '0' }, {
        // '1', '1', '0', '0', '0' },
        // { '0', '0', '1', '0', '0' }, { '0', '0', '0', '1', '1' } }));
        System.out.println(c.numIslands(new int[][] {{0, 0, 0, 1, 0, 0, 0, 1, 1, 0}, {0, 0, 0, 0, 1, 1, 1, 1, 1, 0}, {1, 1, 1, 1, 1, 1, 0, 0, 1, 1}, {0, 0, 1, 1, 0, 1, 0, 0, 0, 1}, {1, 0, 1, 1, 0, 1, 1, 1, 0, 1}, {0, 0, 0, 1, 1, 1, 0, 0, 0, 1}, {1, 1, 1, 0, 1, 1, 1, 0, 1, 1}, {1, 1, 0, 1, 1, 1, 0, 1, 0, 1}, {0, 0, 1, 1, 0, 1, 0, 0, 1, 1}, {0, 1, 1, 0, 1, 0, 0, 0, 1, 1}}));
    }
}

File: src/leetcode/third/Code_200_NumberOfIslands.py
class Solution:
    def numIslands(self, grid):
        if len(grid) == 0:
            return 0
        res = 0
        for i in range(len(grid)):
            for j in range(0, len(grid[0])):
                if grid[i][j] == '1':
                    self.helper(grid, i, j)
                    res += 1
        return res

    def helper(self, grid, i, j):
        if i >= len(grid) or i < 0 or j >= len(grid[0]) or j < 0 or grid[i][j] != '1':
            return
        grid[i][j] = '2'
        self.helper(grid, i + 1, j)
        self.helper(grid, i, j + 1)
        self.helper(grid, i - 1, j)
        self.helper(grid, i, j - 1)


if __name__ == "__main__":
    grid = [['1', '1', '0'], ['1', '1', '0']]
    print(len(grid))
    print(Solution().numIslands(grid))
